File: tools/dogfood/avatar-follower.swift
// avatar-follower.swift — Click-following orchestrator for the avatar overlay.
//
// Adapted from the ball experiment (packages/heads-up/examples/click-follower.swift).
// Key lesson: CGEventTap in .listenOnly mode does NOT block clicks —
// clicks pass through to apps/desktop AND we receive them.
//
// Build:  swiftc -O -o avatar-follower avatar-follower.swift
// Run:    ./avatar-follower   (after avatar canvas exists)

import Foundation
import CoreGraphics

// -- Configuration --
let socketPath = NSString(string: "~/.config/heads-up/sock").expandingTildeInPath
let canvasID   = "avatar"
let canvasSize: Double = 300
let animDuration: Double = 1.2
let animFPS: Double = 60.0

// -- State --
var curX: Double = 700
var curY: Double = 1300
var moveID: UInt64 = 0

// -- Easing (same as ball experiment) --
func easeInOutCubic(_ t: Double) -> Double {
    t < 0.5 ? 4*t*t*t : 1 - pow(-2*t+2, 3)/2
}

// -- Unix socket helpers (from ball experiment) --
func connectSock() -> Int32 {
    let fd = socket(AF_UNIX, SOCK_STREAM, 0)
    guard fd >= 0 else { return -1 }
    var addr = sockaddr_un()
    addr.sun_family = sa_family_t(AF_UNIX)
    let pathBytes = socketPath.utf8CString
    let maxLen = MemoryLayout.size(ofValue: addr.sun_path)
    withUnsafeMutablePointer(to: &addr.sun_path) { ptr in
        pathBytes.withUnsafeBufferPointer { src in
            UnsafeMutableRawPointer(ptr).copyMemory(
                from: src.baseAddress!, byteCount: min(pathBytes.count, maxLen))
        }
    }
    let r = withUnsafePointer(to: &addr) { p in
        p.withMemoryRebound(to: sockaddr.self, capacity: 1) {
            connect(fd, $0, socklen_t(MemoryLayout<sockaddr_un>.size))
        }
    }
    guard r == 0 else { close(fd); return -1 }
    return fd
}

func sendJSON(_ fd: Int32, _ json: String) {
    let line = json + "\n"
    line.withCString { ptr in _ = write(fd, ptr, strlen(ptr)) }
    var buf = [UInt8](repeating: 0, count: 512)
    _ = read(fd, &buf, buf.count)
}

// -- Animation (from ball experiment, adapted for avatar) --
func animateTo(_ clickX: Double, _ clickY: Double, _ mid: UInt64) {
    let sx = curX, sy = curY
    // Center the canvas on the click point
    let tx = clickX - canvasSize / 2
    let ty = clickY - canvasSize / 2

    let fd = connectSock()
    guard fd >= 0 else { return }
    defer { close(fd) }

    let n = Int(animFPS * animDuration)
    let t0 = Date()
    for i in 0...n {
        guard moveID == mid else { break }
        let t = Double(i) / Double(n)
        let e = easeInOutCubic(t)
        let x = sx + (tx - sx) * e
        let y = sy + (ty - sy) * e
        curX = x; curY = y
        sendJSON(fd, "{\"action\":\"update\",\"id\":\"\(canvasID)\",\"at\":[\(x),\(y),\(canvasSize),\(canvasSize)]}")
        let want = Double(i + 1) / animFPS
        let got = Date().timeIntervalSince(t0)
        if want > got { Thread.sleep(forTimeInterval: want - got) }
    }
}

// -- Event tap callback --
var eventTap: CFMachPort? = nil

func tapCB(_ proxy: CGEventTapProxy, _ type: CGEventType, _ event: CGEvent,
           _ refcon: UnsafeMutableRawPointer?) -> Unmanaged<CGEvent>? {
    // macOS disables the tap after certain UI events (page reloads, etc.)
    // Re-enable it immediately when that happens.
    if type == .tapDisabledByTimeout || type == .tapDisabledByUserInput {
        if let tap = eventTap {
            CGEvent.tapEnable(tap: tap, enable: true)
            FileHandle.standardError.write("Event tap re-enabled.\n".data(using: .utf8)!)
        }
        return Unmanaged.passUnretained(event)
    }

    if type == .leftMouseDown {
        let p = event.location
        moveID &+= 1
        let mid = moveID
        DispatchQueue.global(qos: .userInteractive).async {
            animateTo(p.x, p.y, mid)
        }
    }
    return Unmanaged.passUnretained(event)
}

// -- Main --
// Listen for clicks AND tap-disabled events so we can re-enable
let mask = CGEventMask(1 << CGEventType.leftMouseDown.rawValue)
           | CGEventMask(1 << CGEventType.tapDisabledByTimeout.rawValue)
           | CGEventMask(1 << CGEventType.tapDisabledByUserInput.rawValue)

guard let tap = CGEvent.tapCreate(
    tap: .cgSessionEventTap, place: .headInsertEventTap, options: .listenOnly,
    eventsOfInterest: mask,
    callback: tapCB, userInfo: nil
) else {
    FileHandle.standardError.write("Event tap failed. Grant Accessibility permission.\n".data(using: .utf8)!)
    exit(1)
}
eventTap = tap  // Store for re-enable in callback
let src = CFMachPortCreateRunLoopSource(nil, tap, 0)!
CFRunLoopAddSource(CFRunLoopGetCurrent(), src, .commonModes)
CGEvent.tapEnable(tap: tap, enable: true)
FileHandle.standardError.write("""
    avatar-follower running.
      Click anywhere → avatar follows (1.2s ease-in-out-cubic)
      Clicks pass through to apps (listen-only tap).
      Ctrl+C to stop.
    \n
    """.data(using: .utf8)!)
CFRunLoopRun()

File: tools/dogfood/avatar-follower.py
#!/usr/bin/env python3
"""avatar-follower.py — Click-following orchestrator for the avatar overlay.

Monitors global mouse clicks via CGEventTap (listen-only).
Animates the avatar canvas to wherever the user clicks.
Clicks pass through to apps — this is listen-only, not intercepting.

Requires:
  - Accessibility permission (for CGEventTap)
  - heads-up daemon running with avatar canvas created
  - pip3 install pyobjc-framework-Quartz (for CGEventTap)

Usage:
  python3 avatar-follower.py

Architecture lesson from the ball experiment:
  - CGEventTap in kCGEventTapOptionListenOnly does NOT block clicks
  - Clicks go through to the desktop/app AND we receive them
  - This is the "broken hands" design: the agent perceives without interfering
"""

import subprocess
import sys
import time
import threading
import json
import math
import os

HEADS_UP = os.path.join(os.path.dirname(os.path.dirname(os.path.dirname(os.path.abspath(__file__)))),
                         "packages", "heads-up", "heads-up")
CANVAS_ID = "avatar"
CANVAS_SIZE = 300  # matches --at w,h from creation

# Current position (CG coordinates, top-left origin)
cur_x = 700.0
cur_y = 1300.0

# Animation state
move_id = 0
move_lock = threading.Lock()

def ease_in_out_cubic(t):
    if t < 0.5:
        return 4 * t * t * t
    return 1 - pow(-2 * t + 2, 3) / 2

def send_update(x, y):
    """Update avatar canvas position via heads-up."""
    try:
        subprocess.run(
            [HEADS_UP, "update", "--id", CANVAS_ID,
             "--at", f"{x},{y},{CANVAS_SIZE},{CANVAS_SIZE}"],
            capture_output=True, timeout=2
        )
    except Exception:
        pass

def animate_to(tx, ty, mid):
    """Animate avatar to target position with easing."""
    global cur_x, cur_y, move_id

    sx, sy = cur_x, cur_y
    # Center the canvas on the click point
    tx -= CANVAS_SIZE / 2
    ty -= CANVAS_SIZE / 2

    duration = 1.2  # seconds
    fps = 60
    n = int(fps * duration)
    t0 = time.time()

    for i in range(n + 1):
        with move_lock:
            if move_id != mid:
                return  # animation was superseded

        t = i / n
        e = ease_in_out_cubic(t)
        x = sx + (tx - sx) * e
        y = sy + (ty - sy) * e
        cur_x, cur_y = x, y

        send_update(x, y)

        # Frame timing
        want = (i + 1) / fps
        got = time.time() - t0
        if want > got:
            time.sleep(want - got)

def on_click(x, y):
    """Handle a click event — start animation to click position."""
    global move_id

    with move_lock:
        move_id += 1
        mid = move_id

    threading.Thread(target=animate_to, args=(x, y, mid), daemon=True).start()

def main():
    # Try to use CGEventTap via pyobjc
    try:
        from Quartz import (
            CGEventTapCreate, CGEventTapEnable,
            CGEventGetLocation, CGEventMaskBit,
            kCGSessionEventTap, kCGHeadInsertEventTap,
            kCGEventTapOptionListenOnly, kCGEventLeftMouseDown,
            CFMachPortCreateRunLoopSource, CFRunLoopGetCurrent,
            CFRunLoopAddSource, CFRunLoopRun, kCFRunLoopCommonModes
        )

        def callback(proxy, type_, event, refcon):
            loc = CGEventGetLocation(event)
            on_click(loc.x, loc.y)
            return event

        mask = CGEventMaskBit(kCGEventLeftMouseDown)
        tap = CGEventTapCreate(
            kCGSessionEventTap,
            kCGHeadInsertEventTap,
            kCGEventTapOptionListenOnly,
            mask,
            callback,
            None
        )

        if tap is None:
            print("CGEventTap failed. Grant Accessibility permission.", file=sys.stderr)
            sys.exit(1)

        source = CFMachPortCreateRunLoopSource(None, tap, 0)
        CFRunLoopAddSource(CFRunLoopGetCurrent(), source, kCFRunLoopCommonModes)
        CGEventTapEnable(tap, True)

        print("avatar-follower running (CGEventTap).", file=sys.stderr)
        print("  Click anywhere → avatar follows (1.2s ease)", file=sys.stderr)
        print("  Clicks pass through to apps.", file=sys.stderr)
        print("  Ctrl+C to stop.", file=sys.stderr)

        CFRunLoopRun()

    except ImportError:
        print("pyobjc-framework-Quartz not found.", file=sys.stderr)
        print("Install: pip3 install pyobjc-framework-Quartz", file=sys.stderr)
        print("Falling back to polling mode (no click detection).", file=sys.stderr)
        # Fallback: just keep the process alive
        try:
            while True:
                time.sleep(1)
        except KeyboardInterrupt:
            pass

if __name__ == "__main__":
    main()
